src/api/artifact/controller.go

```diff
@@ -54,10 +60,220 @@ type Controller interface {
+			"repository_id": repository.RepositoryID,
+			"digest":        digest,
+		},
+	}
+	_, artifacts, err := c.artMgr.List(ctx, query)
+	if err != nil {
+		return nil, err
+	}
+	// the artifact already exists under the repository, return directly
+	if len(artifacts) > 0 {
+		return artifacts[0], nil
+	}
+
+	// the artifact doesn't exist under the repository, create it first
+	artifact := &artifact.Artifact{
+		ProjectID:    repository.ProjectID,
+		RepositoryID: repository.RepositoryID,
+		Digest:       digest,
+		PushTime:     time.Now(),
+	}
+	// abstract the specific information for the artifact
+	c.abstract(ctx, repository, digest, artifact)
+	// create it
+	id, err := c.artMgr.Create(ctx, artifact)
+	if err != nil {
+		return nil, err
+	}
+	artifact.ID = id
+	return artifact, nil
+}
+
+func (c *controller) ensureTag(ctx context.Context, artifact *artifact.Artifact, name string) error {
+	query := &q.Query{
+		Keywords: map[string]interface{}{
+			"repository_id": artifact.RepositoryID,
+			"name":          name,
+		},
+	}
+	_, tags, err := c.tagMgr.List(ctx, query)
+	if err != nil {
+		return err
+	}
+	// the tag already exists under the repository
+	if len(tags) > 0 {
+		tag := tags[0]
+		// the tag already exists under the repository and is attached to the artifact, return directly
+		if tag.ArtifactID == artifact.ID {
+			return nil
+		}
+		// the tag exists under the repository, but it is attached to other artifact
+		// update it to point to the provided artifact
+		tag.ArtifactID = artifact.ID
+		tag.PushTime = time.Now()
+		return c.tagMgr.Update(ctx, tag, "ArtifactID", "PushTime")
+	}
+	// the tag doesn't exist under the repository, create it
+	_, err = c.tagMgr.Create(ctx, &tag_model.Tag{
+		RepositoryID: artifact.RepositoryID,
+		ArtifactID:   artifact.ID,
+		Name:         name,
+		PushTime:     time.Now(),
+	})
+	return err
+}
+
+func (c *controller) List(ctx context.Context, query *q.Query, option *Option) (int64, []*Artifact, error) {
+	total, arts, err := c.artMgr.List(ctx, query)
+	if err != nil {
+		return 0, nil, err
+	}
+	var artifacts []*Artifact
+	for _, art := range arts {
+		artifacts = append(artifacts, c.assembleArtifact(ctx, art, option))
+	}
+	return total, artifacts, nil
+}
+func (c *controller) Get(ctx context.Context, id int64, option *Option) (*Artifact, error) {
+	art, err := c.artMgr.Get(ctx, id)
+	if err != nil {
+		return nil, err
+	}
+	return c.assembleArtifact(ctx, art, option), nil
+}
+
+func (c *controller) Delete(ctx context.Context, id int64) error {
+	// delete artifact first in case the artifact is referenced by other artifact
+	if err := c.artMgr.Delete(ctx, id); err != nil {
+		return err
+	}
+
+	// delete all tags that attached to the artifact
+	_, tags, err := c.tagMgr.List(ctx, &q.Query{
+		Keywords: map[string]interface{}{
+			"artifact_id": id,
+		},
+	})
+	if err != nil {
+		return err
+	}
+	for _, tag := range tags {
+		if err = c.DeleteTag(ctx, tag.ID); err != nil {
+			return err
```

Comment:
This will be put in a transaction after weiwei's PR merged.

Reply:
delete tags firstly, then to delete af?

Reply:
This is explained by comment: `delete artifact first in case the artifact is referenced by other artifact`.
When calling `c.artMgr.Delete`, artifact manager checks whether the artifact can be deleted first(when the artifact is referenced by others, it cannot be deleted).